Create AGENTS.md (#12210)
JuliaPackaging/Yggdrasil: AGENTS.md

# Agent Guide for Yggdrasil Build Scripts

This guide helps AI agents generate correct `build_tarballs.jl` recipes for BinaryBuilder.jl in the Yggdrasil repository.

## Prerequisites

> - **BinaryBuilder.jl**: Requires Julia 1.3.0 or later

Reviewer: At this point, I would say recommend only using 1.7 - I don't see a reason to use an earlier version given how far out of date we are already.

> - **BinaryBuilder.jl**: Requires Julia 1.3.0 or later

Author: There's one remaining issue @giordano pointed out with using 1.12+. The goal is still to have that working.. so yeah pin to 1.7 for now and update soon™️

> - **BinaryBuilder.jl**: Requires Julia 1.3.0 or later

Reviewer 2: Yes, I'm trying to reduce it right now.

- Use `supported_platforms()` to get all available platforms

## Essential Structure

Every `build_tarballs.jl` file follows this pattern:

```julia
using BinaryBuilder

name = "PackageName"              # Valid Julia identifier (no spaces/dashes/dots)
version = v"X.Y.Z"                # Only major.minor.patch (no prerelease/build tags)

sources = [
    ArchiveSource("URL", "sha256hash"),
    # GitSource("URL", "commit_hash"),
    # DirectorySource("./bundled"),  # for patches
]

script = raw"""
cd ${WORKSPACE}/srcdir/package-*
# Build commands here
"""

platforms = supported_platforms()  # or filter as needed

products = [
    LibraryProduct("libname", :libname),
    ExecutableProduct("exename", :exename),
]

dependencies = [
    Dependency("SomeDep_jll"),
    # BuildDependency("BuildOnlyDep_jll"),
]

build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies)
```

## Critical Rules

### Naming

- **Name**: Must be a valid Julia identifier. Replace spaces/dashes with underscores. Match upstream casing.
- **Version**: Only `X.Y.Z` format. Truncate any `-alpha`, `+build`, or 4+ level versions.

> - **Products**: Export symbols match the library/executable names (as symbols: `:libname`).

Reviewer: This isn't a hard rule actually, just general guidance. It may be that it makes more sense to name it something else.

### Sources

- **ArchiveSource**: For tarballs (`.tar.gz`, `.tar.xz`, `.zip`). Always include SHA256 hash.
- **GitSource**: For git repos. Use specific commit hash, not branch names.
- **DirectorySource**: For local patches. Place patches in `bundled/patches/` subdirectory.
- Build **one package per recipe**. Don't bundle multiple packages—use dependencies instead.

### Build Script

The script runs in `x86_64-linux-musl` environment. Key variables:

- `${prefix}`: Install root (target for all outputs)
- `${bindir}`: Executables go here (= `${prefix}/bin`)
- `${libdir}`: Libraries go here (= `${prefix}/bin` on Windows, `${prefix}/lib` elsewhere)
- `${includedir}`: Headers go here (= `${prefix}/include`)
- `${WORKSPACE}/srcdir`: Where sources are extracted
- `${target}`: Target platform triplet
- `${nproc}`: Number of parallel jobs
- `${CC}`, `${CXX}`, `${FC}`: Cross-compilers for C/C++/Fortran

### Common Build Systems

**Autotools**:

```bash
./configure --prefix=${prefix} --build=${MACHTYPE} --host=${target}
make -j${nproc}
make install
```

> Use `update_configure_scripts` first if `config.sub` is old.

Reviewer: Does it need to know what "old" means here?

```bash
cmake -B build \
    -DCMAKE_INSTALL_PREFIX=${prefix} \
    -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} \
    -DCMAKE_BUILD_TYPE=Release
cmake --build build --parallel ${nproc}
cmake --install build
```

**Meson**:

```bash
meson setup build --cross-file="${MESON_TARGET_TOOLCHAIN}"
meson compile -C build -j${nproc}
meson install -C build
```

**Make**:

```bash
make -j${nproc} PREFIX=${prefix}
make install PREFIX=${prefix}
```

**Go** (add `compilers=[:c, :go]`):

```bash
go build -o ${bindir}/executable
```

**Rust** (add `compilers=[:c, :rust]`):

```bash
cargo build --release
install -Dm755 target/*/release/executable ${bindir}/executable
```

### Platform-Specific Logic

```bash
if [[ "${target}" == *-mingw* ]]; then
    # Windows-specific
elif [[ "${target}" == *-apple-* ]]; then
    # macOS-specific
elif [[ "${target}" == *-freebsd* ]]; then
    # FreeBSD-specific
fi

if [[ ${nbits} == 32 ]]; then
    # 32-bit specific
fi
```

### Products

- **LibraryProduct**: Shared libraries (`.so`, `.dylib`, `.dll`)
- **ExecutableProduct**: Binary executables
- **FileProduct**: Other files (headers, data files)
- **FrameworkProduct**: macOS frameworks

### Dependencies

> - **Dependency**: Runtime dependency (will be in JLL package)

Reviewer: ```suggestion
- **Dependency**: Runtime dependency (will be a dependency of the generated JLL package)
```

> - **BuildDependency**: Build-time only (not in final JLL)

Reviewer: ```suggestion
- **BuildDependency**: Build-time only (not a dependency for the final JLL)
```

> - **HostBuildDependency**: Runs on build host, not target

Reviewer: ```suggestion
- **HostBuildDependency**: Build-time only dependency that needs to run on the build host, not target (not a dependency for the final JLL)
```

Always add `_jll` suffix: `Dependency("Zlib_jll")`

> ### GCC Version Selection

Reviewer: Can we link other documents here? It needs to be one of the versions we actually have built, so basically only things in this table: https://github.com/JuliaPackaging/Yggdrasil/blob/master/RootFS.md#compiler-shards

Use `preferred_gcc_version=v"X"` for:

- **C++ code**: Use oldest GCC that compiles (≤10 for Julia v1.6 compatibility)
- **Dependencies built with newer GCC**: Match or exceed their GCC version
- **Musl bugs**: Use GCC ≥6 to avoid `posix_memalign` issues
- Default is GCC 4.8.5 for maximum compatibility

### Optional Arguments

```julia
build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies;
    julia_compat="1.10",                   # Minimum Julia version
    preferred_gcc_version=v"8",            # GCC version
    preferred_llvm_version=v"13",          # LLVM version
    compilers=[:c, :rust],                 # Additional compilers
    clang_use_lld=false,                   # Use LLD linker
)
```

## Common Patterns

### Applying Patches

```julia
sources = [
    ArchiveSource("..."),
    DirectorySource("./bundled"),
]

script = raw"""
cd ${WORKSPACE}/srcdir/package-*
atomic_patch -p1 ${WORKSPACE}/srcdir/patches/fix.patch
# ... rest of build
"""
```

### Platform Filtering

```julia
# Only 64-bit platforms

> platforms = filter(p -> arch(p) != "i686" && arch(p) != "armv7l", supported_platforms())

Reviewer: ```suggestion
platforms = filter(p -> nbits(p) == 64, supported_platforms())
```
This is better. Also, the original one was missing the `armv6l` from the filter.

> # Expand C++ string ABI variants
> platforms = expand_cxxstring_abis(platforms)

Reviewer: ```suggestion
# Expand C++ string ABI variants
platforms = expand_cxxstring_abis(platforms)

# Expand Fortran library versions (only for Fortran codes)
platforms = expand_gfortran_versions(platforms)
```
If we talk about the C++ expansion, we should also mention the Fortran expansion that is needed.

# Specific platforms
platforms = [Platform("x86_64", "linux"), Platform("x86_64", "macos")]
```

### Setting Environment Variables

```bash
export CPPFLAGS="-I${includedir}"
export LDFLAGS="-L${libdir}"
export PKG_CONFIG_PATH="${prefix}/lib/pkgconfig:${PKG_CONFIG_PATH}"
```

### Multiple Source Directories

```bash
cd ${WORKSPACE}/srcdir
mkdir build && cd build
cmake ../package-source

> ```

Reviewer: I am confused what this means by "multiple source directories" - this just looks like a separate build directory for Cmake.

Author: Changed to `Out-of-Source Builds`. Is that clearer?

```

Place in alphabetical directory: `X/XYZ/build_tarballs.jl`

## Commit Messages

Format: `[PackageName] Brief description`

Examples:

- `[Zlib] Update to v1.3.1`
- `[CMake] Add OpenSSL dependency`
- `[FFMPEG] Fix build on FreeBSD`

Use `[skip build]` to publish JLL without rebuilding (for metadata-only changes like compat bounds).

## Testing JLL Packages Before Merging

Before merging a Yggdrasil PR, you should test the generated JLL package locally to ensure it works correctly. This avoids automatic registration of broken packages in the General registry.

### Step 0: Speed Up Testing (Optional)

For faster local testing, build only for your current platform by filtering the platforms list. Add this line after the `platforms = [...]` definition:

```julia
# For local testing, only build for the current platform
platforms = filter(p -> Sys.isapple() && arch(p) == "aarch64", platforms)  # macOS ARM64
# platforms = filter(p -> Sys.isapple() && arch(p) == "x86_64", platforms)  # macOS x86_64
# platforms = filter(p -> Sys.islinux() && arch(p) == "x86_64", platforms)  # Linux x86_64
# platforms = filter(p -> Sys.iswindows() && arch(p) == "x86_64", platforms)  # Windows x86_64

> ```

Reviewer: IMO, it is better just to pass your local platform to the build call, e.g.,
```
julia --project=/path/to/Yggdrasil build_tarballs.jl --verbose --debug x86_64-linux-gnu
```

instead of doing a lot of these steps.

# Agent Guide for Yggdrasil Build Scripts

This guide helps AI agents generate correct `build_tarballs.jl` recipes for BinaryBuilder.jl in the Yggdrasil repository.

## Prerequisites

- **BinaryBuilder.jl**: Requires Julia 1.7 specifically (use `julia +1.7` if juliaup is installed)
- **Supported Platforms**: Linux (glibc and musl for x86_64, i686, aarch64, armv7l, armv6l, ppc64le, riscv64), Windows (x86_64, i686), macOS (x86_64, aarch64), FreeBSD (x86_64, aarch64)
- Use `supported_platforms()` to get all available platforms

## Special Dependencies

Some dependencies require special handling:

- **LLVM packages**: Must use `LLVM_full_jll` and match the version used by the Julia version. Requires careful ABI compatibility.
- **MPI packages**: Need `MPIPreferences.jl` configuration and must use `MPItrampoline_jll` for cross-implementation compatibility.
- **CUDA packages**: Require `CUDA_Runtime_jll` and must handle different CUDA versions. GPU code needs special compilation flags.

> For these complex dependencies, consult existing recipes in the repository (search for `LLVM_full_jll`, `MPItrampoline_jll`, or `CUDA_Runtime_jll`).

Reviewer: Actually, for CUDA the packages probably don't have `CUDA_Runtime_jll` in their file, but instead they call `CUDA.required_dependencies` to get that.
